crates/maybenot/src/time.rs

```diff
@@ -0,0 +1,61 @@
+use std::ops::AddAssign;
+
+/// Trait representing instants in time. Allows using maybenot frameworks with
+/// custom time sources. If you want to use maybenot with a different time source
+/// than `std::time::Instant`, implement this trait for your instant type, and the
+/// [`Duration`] trait for your corresponding duration type.
+pub trait Instant: Clone + Copy {
+    type Duration: Duration;
+
+    /// Returns the amount of time elapsed from another instant to this one.
+    ///
+    /// Should return a zero duration if `earlier` is later than `self`
+    fn duration_since(&self, earlier: Self) -> Self::Duration;
+}
+
+pub trait Duration: Clone + Copy + AddAssign + PartialOrd {
+    /// Creates a new duration, spanning no time.
+    fn zero() -> Self;
+
+    /// Creates a new duration from the specified number of microseconds.
+    fn from_micros(micros: u64) -> Self;
+
+    /// Returns the total number of whole microseconds contained by this duration.
+    fn as_micros(&self) -> u128;
+
```

Comment:
Observation: For some native `Duration` representations, conversion to and from microseconds is not a cheap operation.

For example, if you're using `mach_timebase_info` or `QueryPerformanceCounter`, the native representation is an a system-dependent unit, and you need to multiply and divide by a run-time-queried numerator and denominator in order to get nanoseconds or microseconds.

Or more prosaically, if the representation is in nanoseconds, you need to divide by 1000.

This wouldn't be a big concern, except that division is fairly expensive when you're doing it on types that are wider than the CPU supports (e.g, 64-bit division is expensive on most 32-bit CPUs).

Because of these performance concerns, it's often better - when we need to treat Durations as numbers for the purposes of computation - to compute with a unit of "ticks", where a "tick" is an arbitrary unit that the Duration considers easiest to compute with.  

To implement this, you would add a from_ticks() and as_ticks() methods to return the Duration in units of "ticks", and then you'd use "ticks" for your computations internally.  

----

This is a change you could add later, in a backward-compatible way, later on, if you want: you would introduce default implementations of `from_ticks` and `as_ticks` that just used microseconds for ticks by default.  

But on the other hand, I think you could just rename these functions now: I don't see any place in the code where you are relying on their unit being "microseconds" instead of "some arbitrary small unit of time".

Reply:
I just did what `maybenot` does currently. A lot of the internal timing stuff is microsecond based. However I don't know if they *have to be*. I'll let @pylls answer that.

Reply:
Thanks for the detailed dive into this. When I have time I can look at how simple it would be to maybe be agnostic over time unit. But I don't have that deep understanding of details in the framework. So I still need feedback from people who do.

Reply:
The bad news: A fundamental part of the framework is interpreting sampled distributions (`f64`) as microseconds and turning them into durations. We can't dynamically adjust this part: we should lock at microseconds or nanoseconds and make sure the choice here is reasonable long-term and can be efficiently implemented.

The good news: I think we can eliminate all `as_micros` by revising how we treat blocking limits in machines. Currently, the framework tracks blocking duration (as `Duration`) but goes to/from micros due to the `allowed_blocked_microsec: u64` limit in each machine. That this is an `u64` is an artefact from how I added blocking limits analogous to padding limits (`allowed_padding_packets: u64`). We can simply have this limit be a `Duration` and we're good, at least for Maybenot v2.

Would that help?

Reply:
Edited the post above a couple of times, then went down the rabbit hole of distributions, so a new post instead.

Actually [randr_dist](https://crates.io/crates/rand_distr) uses [trait num_traits::float::Float](https://docs.rs/num-traits/0.2.19/num_traits/float/trait.Float.html) for all floats. For most dists (some return `u64`), it would be possible to go for the Float trait.

The above could be used to get more suitable types for 32-bit CPUs, but, we would still have to pick what a sampled distribution means: is it microseconds, nanoseconds, etc, and turn it into a Duration. This is fundamental.

In m opinion: for v2, get rid of `as_micros` for one of the options above, then go with this branch for now. If we learn that this matters in a while, great, I'm not against significant efforts if we can justify it and we'll get it into v3. Thoughts, @faern and @nmathewson ?

Reply:
Ok, yeah you are right. I read how `as_micros` is used more carefully now. As you say, it's only ever used to actually compute the fraction between two durations. I'll look into changing the trait to only require it to provide division with itself resulting in an `f64`

Reply:
This turned out very nicely! Simple implementation for the default case of `std::time::Duration`, and is flexible with how other implementers decide to represent themselves as a float. Since it's a float ratio we *need* one float division is unavoidable, but with this implementation it should be possible to reduce them at least.

Reply:
Nice! I pushed a small update to get rid of one more `from_micros()` call. To me, this is looking worth to merge now.

Reply:
Nice! Yeah that looks good. The only things I think we have not proven, that could be useful before merging:
* Is there any performance difference in the library while using `std::time` before and after this PR? We don't want to degrade performance since ultimately this entire PR is only for increasing possible performance.
* Can the significant speedup of using `coarsetime` still be shown with this version of the framework? If the time abstraction for some reason prevents that speedup from happening, the abstraction also becomes pointless.